File: src/trie.js

var Node = function() {
  this.next = {};
}

Node.prototype.insert = function (word) {
  
}

var Trie = function() {
  this.root = new Node;
}

Trie.prototype.insert = function (word) {
  var node = this.root;
  var i = 0;
  while (i < word.length) {
    var curChar = word.charAt(i);
    if (!node.next.hasOwnProperty(curChar)) {
      node.next[curChar] = new Node;
    }
    node = node.next[curChar];
    i += 1;
  }
}

module.exports = Trie;

File: src/trie.cc
#include <cstdio>
#include <string>
#include <map>

using namespace std;

struct Node {
  map<char, Node*> next;
  static int count;
  Node() {
    ++count;
  }
};
int Node::count = 0;

struct Trie {
  Node* root;
  Trie(): root(new Node) {}

  void insert(const string& s) {
    Node* node = root;
    for(int i = 0; i < s.size(); ++i) {
      if (node->next.find(s[i]) == node->next.end()) {
        node->next[s[i]] = new Node;
      }
      node = node->next[s[i]];
    }
  }
};

int main() {
  char s[100];
  Trie t;
  for (int i = 0; i < 100000; ++i) {
    for (int j = 0; j < 20; ++j) {
      s[j] = rand() % 128;
    }
    s[20] = 0;
    t.insert(s);
  }
  printf("%d nodes created\n", Node::count);
}
